geniza/corpus/migrations/0034_merge_duplicate_tags.py

```diff
@@ -0,0 +1,88 @@
+# Generated by Django 3.2.14 on 2022-08-08 17:50
+
+from django.conf import settings
+from django.contrib.admin.models import CHANGE
+from django.db import migrations
+from unidecode import unidecode
+
+
+def lowercase_and_unidecode_tags(apps, schema_editor):
+    """
+    lowercase all tags, merge lowercase and uppercase tags into lowercase;
+    remove diacritics from all tags, merge tags that are the same except for diacritics
+    """
+    ContentType = apps.get_model("contenttypes", "ContentType")
+    Document = apps.get_model("corpus", "Document")
+    LogEntry = apps.get_model("admin", "LogEntry")
+    Tag = apps.get_model("taggit", "Tag")
+    TaggedItem = apps.get_model("taggit", "TaggedItem")
+    User = apps.get_model("auth", "User")
+
+    doc_contenttype = ContentType.objects.get(app_label="corpus", model="document")
+    tag_contenttype = ContentType.objects.get(app_label="taggit", model="tag")
+    (user, _) = User.objects.get_or_create(
+        username=settings.SCRIPT_USERNAME, is_staff=False, is_active=False
+    )  # generic user for script execution
+    tags_to_delete = []
+    tagged_items_to_delete = []
+
+    for tag in Tag.objects.all():
+        lower_ascii_name = unidecode(tag.name).lower()
```

Comment:
I would prefer not to _force_ lower case, but just to use the taggit case insensitive logic to avoid case-collisions. We should preserve the mixed-case tag where making case-insensitive results in duplicates. So logic instead should be something like:

- check if `unidecode` changes the current tag; use that as basis for new tag if changed
- do a case-insensitive search for any tags that should be merged with the current one; choose the mixed case variant if there is one
- for any tag that is being removed due to collapse, do a tagged items queryset update to replace the old tag id with the merged tag id (should not need to change anything else)
- remove the redundant tags

It might be easier to aggregate the tags first rather than iterating on`Tag.objects.all()`; then you don't have to worry about what order you see them in. But perhaps it is sufficient to sort by tag name, since you should get the upper case variants first.

Reply:
@rlskoeser I made good progress on this, but I was wondering if you could help me with the aggregation query. The problem I'm running into when iterating on `Tag.objects.all()` is that, if there is more than one variant of a tag name in terms of diacritics, the ones alphabetically after the first one aren't picked up for merge. Maybe it's just a problem with my query for finding the tags to merge:

```python
for tag in Tag.objects.all().order_by("name"):
    if tag.pk not in tags_to_delete:
        ascii_name = unidecode(tag.name)

        # search for any differently-cased (incl. unidecoded) versions of this tag
        tags_to_merge = Tag.objects.filter(
            Q(name__iexact=ascii_name) | Q(name__iexact=tag.name)
        ).exclude(pk=tag.pk)
```

That's only going to pick up other tags with the exact same diacritics, or tags that are the same but in ASCII. If the first encountered tag does not have diacritics, but the second does, then it won't pick up the second one.

The problem appears when I then go to save the tag with its unidecoded name, leading to a uniqueness violation when the second variant is reached (in the data I have, "Abu Sahl Levi" is picked up first, then "Abū Sahl Levi" is the second variant):
```python
        # rename this tag to unidecoded name
        if ascii_name != tag.name:
            tag.name = ascii_name
            tag.save()
```

My idea was to skip over all the merged tags in the loop by adding them to `tags_to_delete`:

```python
        # merge all other matched tags into this one
        for matched_tag in tags_to_merge:
            # replace the matched tag in all tagged items with this tag
            matched_tag.taggit_taggeditem_items.update(tag=tag)
            # finally, mark matched tag for deletion
            tags_to_delete.append(matched_tag.pk)
```

In this case "Abū Sahl Levi" never gets added to `tags_to_delete` because it wasn't picked up in the merge queryset.

Is there a way we can write an aggregation query that will handle all diacritic variants of a tag?

Reply:
I think it might be easier to do the aggregation first in python! That's what I meant when I said "aggregate the tags first" but I see now that isn't clear. Let me take a quick stab at this.